compilers/concrete-compiler/compiler/lib/Dialect/FHELinalg/IR/FHELinalgOps.cpp

```diff
@@ -1687,6 +1687,488 @@ void ToUnsignedOp::getCanonicalizationPatterns(
   getSignedConvCanonicalizationPatterns<ToUnsignedOp>(patterns, context);
 }
 
+std::optional<mlir::Value>
+fuseBackToBackTableLookups(mlir::Value value, mlir::PatternRewriter &rewriter) {
+
+  using mlir::concretelang::FHE::FheIntegerInterface;
+
+  auto currentOperation = value.getDefiningOp();
+
+  auto currentOperationIsTlu =
+      llvm::isa<ApplyLookupTableEintOp>(currentOperation);
+  auto currentOperationIsMappedTlu =
+      llvm::isa<ApplyMappedLookupTableEintOp>(currentOperation);
+
+  if (!currentOperationIsTlu && !currentOperationIsMappedTlu) {
+    return std::nullopt;
+  }
+
+  auto intermediateValue =
+      (currentOperationIsTlu
+           ? (llvm::dyn_cast<ApplyLookupTableEintOp>(currentOperation).getT())
+           : (llvm::dyn_cast<ApplyMappedLookupTableEintOp>(currentOperation)
+                  .getT()));
+
+  auto intermediateOperation = intermediateValue.getDefiningOp();
+  if (!intermediateOperation) {
+    return std::nullopt;
+  }
+
+  auto intermediateOperationIsTlu =
+      llvm::isa<ApplyLookupTableEintOp>(intermediateOperation);
+  auto intermediateOperationIsMappedTlu =
+      llvm::isa<ApplyMappedLookupTableEintOp>(intermediateOperation);
+
+  if (!intermediateOperationIsTlu && !intermediateOperationIsMappedTlu) {
+    return std::nullopt;
+  }
+
+  auto inputValue =
+      (intermediateOperationIsTlu
+           ? (llvm::dyn_cast<ApplyLookupTableEintOp>(intermediateOperation)
+                  .getT())
+           : (llvm::dyn_cast<ApplyMappedLookupTableEintOp>(
+                  intermediateOperation)
+                  .getT()));
+
+  class Indexer {
+  public:
+    int64_t tableSize;
+    bool isSigned;
+
+    Indexer(int64_t tableSize, bool isSigned)
+        : tableSize{tableSize}, isSigned{isSigned} {}
+
+    int64_t sanitizeIndex(int64_t index) const {
```

Comment:
This looks very similar to the calculations done by the `lookup` lambda in [Line 497](https://github.com/zama-ai/concrete/pull/709/commits/6c24f426b86429c3f8b3a9ae78d9ab7bd387a4a5#diff-754013c90529f303dfc1c798d79bb8f856da6b5a9d917e8915dc035ae9502fe6R497). Would it make sense to factor this out?